docs/design.md

```diff
@@ -0,0 +1,153 @@
+For example, a commonly used model loader would be the HuggingFace model loader
+where by providing the model name would load the model to the Model object
+bucket location. Model loaders are defined as container images and can either
+point to prebuilt container images or a Git repository with Dockerfile.
+
+A model
+
+The Model object can optionally describe the training parameters including a reference to the Dataset that should be used for training. In this case, the Model object represents the eventual output of a training/finetuning operation.
+
+```yaml
+kind: Model
+spec:
+  container: # used for either loader or trainer
+  loader: {} # Optional, when empty an just a location will be created with no files in it
+  trainer: {} # Optional, will run a training job
+```
+
+### Model Sources
+
+A Model can be sourced from Git or a prebuilt image. Using Git will trigger a container build Job.
+
+```yaml
+kind: Model
+name: falcon-7b
+spec:
+  container:
+    git:
+      url: https://github.com/substratusai/hf-model-loader
+    image: substratusai/hf-model-loader
+  loader:
+    params: {name: tiuu/falcon-7b}
+    # optional future
+    paramsFrom: configMap?
+```
+
+`params` get converted to environment variables following a specific scheme
+
+The controller will orchestrate the following flow in this case:
+
+<img src="./diagrams/model-building.excalidraw.png" width="80%"></img>
+
+
+### Model Training
+
+Approach 1: Include in Model object
+
+Models can be trained by specifying the `.spec.training` section.
+
+This will orchestrate a training Job with multiple FUSE mounted buckets.
+
+<img src="./diagrams/model-training.excalidraw.png" width="80%"></img>
+
+```yaml
+kind: Model
+name: falcon-7b-k8s
+spec:
+  container:
+  # can only specify git or image not both
+    git: 
+      url: https://github.com/substratusai/hf-llm-trainer
+      ref: refs/branch/main
+    image: substratusai/hf-llm-peft-trainer # eventually introduce native pytorch trainer
+
+  trainer:
+    # sourceModel can be left out for creating new base model
+    sourceModel:
+      name: falcon-7b
+      namespace: other-namespace
+    dataset:
+      name: k8s-instructions
+      namespace: other-namespace
+    resources: {}
+    params: {epochs: 1}
+```
+
+
+
+### Model Training with Notebook
+Notebook container contract is to run notebook.sh and serve
+notebook compatible HTTP endpoint on port 8888.
+
+User flow: Random experiments
+
+notebook-gpu image has transformers, pytorch, cuda, python 3, etc.. basically most you need
+
+example notebook in notebook.
+
+1. Create Notebook with no sourceModel
+2. Load a model dynamically in my notebook session
+3. Experiment with the model and training and only save my notebook somewhere for reference
+
+```yaml
+kind: Notebook
+name: notebooks-for-anything
+spec:
+  container:
+    # or git repo as a source
+    image: substratusai/notebook-gpu
+  resources: {gpuCount: 4, type: L4}
```

Comment:
I think resources should be exactly the same as container resources:
```
    resources:
      requests:
        memory: "10Gi"
        cpu: "10"
      limits:
       nvidia.com/gpu: 2
```
would be valid here.

The open question then would be how to pass it GPU type. Should that be a separate attribute outside of resources?  I vote for yes gpu_type has separate attribute such that we can use standard K8s resource requests which most are familiar with.

Reply:
I am not sure we should directly expose kubernetes `.resources`. Mainly because of what you called out here... the use of specialized equipment like GPUs requires a slew of other fields to be set. I think substratus should simplify this by exposing a few high-level resource knobs. Perhaps:

```yaml
runtime:
  compute: ["GPU", "CPU"] # Can be run on either GPUs or CPUs.
  strategy: Fastest # or "Cheapest", or "Balanced"
  memory: "10Gi" # Translates to AT LEAST 10Gi of gpu-memory and 10Gi or cpu-memory. If more is provisioned (commonly the case), it will be reported to the container using environment variables.
  disk: "50Gi"
```

CPU cores are not really a requirement... more will be faster.

Reply:
When you define a model you have to decide if you want to run the loader on CPU or GPU

Alternative proposal for runtime under loader or trainer section
```
runtime:
  accelerator: # optional if not specified will only use CPU
     type: nvidia-tesla-a100
     count: 8
  cpu: 2 # default will always be 2 if not specified
  memory: "10Gi" # will always be 10 if not specified. Translates to AT LEAST 10Gi of gpu-memory and 10Gi or cpu-memory. If more is provisioned (commonly the case), it will be reported to the container using environment variables.
  disk: "50Gi" # will always be 50Gi if not specified
```

Reply:
When the specify the accelerator in the Model object, the Model object is no longer portable across clusters that have access to different accelerators.

Reply:
I want to specify exactly the resources for my training job. The model loaded in GCS is still portable across clusters. I think I'm misunderstanding the portability concern

Reply:
Perhaps there's a middleground where `runtimeProfile`s can be created by administrators and used by consumers in the form you're pitching @nstogner . Personally, I lean toward starting with the more raw resource spec model since model requirements are quite specific.

Reply:
> I want to specify exactly the resources for my training job. The model loaded in GCS is still portable across clusters. I think I'm misunderstanding the portability concern

Guessing at what Nick meant here: where might models not achieve portability:
1. `resource_limits` might differ from cluster to cluster.
2. different accelerator types are supported in different regions/zones and across different cloud providers.
3. even in the case where both of the above criteria are met, stock out might mean another accelerator type must be used if the job is to proceed.

I think at the moment I still lean toward raw resource specs as @samos123 is pushing for here but potentially a way to think about this is giving an ordered list of hardware specs so if one resource profile fails, another one can be tried.

Reply:
Resolved in slack